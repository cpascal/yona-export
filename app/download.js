import unirest from 'unirest';
import fse from 'fs-extra';
import path from 'path';
import config from '../config';

export const download = (attachment, cb) => {
  let yonaFileUrl = config.YONA.SERVER + '/files/' + attachment.id;

  fse.ensureDirSync(path.join(config.EXPORT_BASE_DIR, config.ATTACHMENTS_DIR, attachment.id +'', '/'));
  unirest.get(yonaFileUrl)
      .headers({
        'Accept': '*/*',
        'Yona-Token': config.YONA.USER_TOKEN
      })
      .end(response => {
        if (response.status !== 200) console.log(response.status);
        if(cb && typeof cb === 'function') {
          cb();
        }
      }).pipe(fse.createWriteStream('./exported/files/' + attachment.id + '/' + attachment.name));
};

