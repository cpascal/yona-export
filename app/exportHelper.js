/* eslint-disable indent */
import path from 'path';

export function getUrlToPost(item, parent, config) {
  if (parent && parent.type) {
    switch (parent.type) {
      case 'ISSUE_POST':
        return config.YONA.TO.SERVER
            + path.join(
                config.YONA.TO.ROOT_CONTEXT,
                '/-_-api/v1/owners/',
                config.YONA.TO.OWNER_NAME,
                '/projects/',
                config.YONA.TO.PROJECT_NAME,
                '/issues',
                parent.number.toString(),
                '/comments');
      case 'BOARD_POST':
        return config.YONA.TO.SERVER
            + path.join(
                config.YONA.TO.ROOT_CONTEXT,
                '/-_-api/v1/owners/',
                config.YONA.TO.OWNER_NAME,
                '/projects/',
                config.YONA.TO.PROJECT_NAME,
                '/posts',
                parent.number.toString(),
                '/comments');
      default:
        throw Error('Unknown item: ', item);
    }
  }

  switch (item.type) {
    case 'ISSUE_POST':
      return config.YONA.TO.SERVER
          + path.join(
              config.YONA.TO.ROOT_CONTEXT,
              '/-_-api/v1/owners/',
              config.YONA.TO.OWNER_NAME,
              '/projects/',
              config.YONA.TO.PROJECT_NAME,
              '/issues');
    case 'BOARD_POST':
      return config.YONA.TO.SERVER
          + path.join(
              config.YONA.TO.ROOT_CONTEXT,
              '/-_-api/v1/owners/',
              config.YONA.TO.OWNER_NAME,
              '/projects/',
              config.YONA.TO.PROJECT_NAME,
              '/posts');
    default:
      throw Error('Unknown item: ', item);
  }
}

export function getItemType(item, parent) {
  const itemType = item.type || parent.type;
  switch (itemType) {
    case 'ISSUE_POST':
      return 'issues';
    case 'BOARD_POST':
      return 'posts';
    default:
      throw Error('Unknown item: ', item);
  }
}
