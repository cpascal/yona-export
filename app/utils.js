import moment from 'moment';

export const getTitle = title => {
  let specialCharsRemoved = title.replace(/[|&;\\\/:.'?*$%@"<>+,()]/g, '').trim();  // cleaning and trimming
  return specialCharsRemoved.split(' ').join('-');
};

export const getDefaultFileName = post => {
  const title = getTitle(post.title);
  const number = lpadZero(post.number, 5) || lpadZero(post.id, 5) || '?????';
  const date = getDateString(post.createdAt || post.due_on || '????');
  const extName = '.md';

  return (`${number}-${date}-${title}${extName}`);
};

export const lpadZero = (no, n) =>
    (no + '').length < n ? lpadZero('0' + no, n) : no;

export const getDateString = (timestamp) => {
  moment.locale('ko-KR');
  let date = moment(new Date(timestamp));
  return date.format('YYYYMMDD');
};
