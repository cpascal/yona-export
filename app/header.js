import matter from 'gray-matter';
import Humanize from 'humanize-plus';
import { download } from './download';

let totalWillDownload = 0, totalDidDownload = 0;
let queuedFiles = {};
const downloadAttachments = function (attachments = []) {
  if (attachments.length > 0) {
    const file = attachments.pop();
    totalWillDownload++;
    console.log(`${totalDidDownload} / ${totalWillDownload} is done. ${file.name} ${Humanize.filesize(file.size)} will be download at ${file.id}.`);
    queuedFiles[file.id] = { name: file.name, size: Humanize.filesize(file.size) };
    download(file, () => {
      totalDidDownload++;
      console.log(`............ ${file.name}: ${Humanize.filesize(file.size)} downloaded.`);
      delete queuedFiles[file.id];
      console.log(`in queued files: `);
      console.dir(queuedFiles);
      downloadAttachments(attachments);
    })
  }
};

export const createHeader = (post) => {
  let header = { ...post };
  if (post.createdAt) {
    header.createdAt = post.createdAt;
  }
  if (post.comments) {
    header.comments = stripKeys(post.comments, ['type']);
    post.comments.forEach(comment => {
      if (comment.attachments) {
        console.log(`${comment.attachments.length} files will be download.`);
        downloadAttachments(comment.attachments);
      }
    });
  }
  if (post.attachments) {
    header.attachments = stripKeysFromAttachment(post.attachments);
    console.log(`${post.attachments.length} files will be download.`);
    downloadAttachments(header.attachments);
  }

  delete header.body;
  return matter.stringify(post.body || '', header, { delims: '```' }).trim();
};

export default {
  createHeader
};

function stripKeysFromAttachment(attachments) {
  return stripKeys(attachments, ['containerId', 'containerType']);
}

function stripKeys(arr = [], keys = []) {
  arr.forEach(item => {
    if (item.attachments) stripKeysFromAttachment(item.attachments);
    keys.forEach(key => {
      delete item[key];
    })
  });
  return arr;
}
