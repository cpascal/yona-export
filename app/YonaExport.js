import unirest from 'unirest';
import path from 'path';
import fse from 'fs-extra';
import config from '../config';
import { replaceAttchementFileId } from './utils';
import { getItemType, getUrlToPost } from './exportHelper';

/**
 * config에 지정된 Yona 서버로 글을 보내는 기능을 담당하는 클래스
 *
 * pushFiles 로 먼저 파일들을 업로드 하고
 * pushPost 로 해당 파일을 첨부파일로 지정해서 글을 옮긴다.
 */
export default class YonaExport {
  constructor() {
    this.okCount = 0;
  }

  pushPost(post, parent, cb) {
    let files = this.collectAttachmentFileIds(post);
    post.body = replaceAttchementFileId(post.body, post.attachments);
    let data = {
      'temporaryUploadFiles': files,
      ...post
    };


    const yonApiUrl = getUrlToPost(post, parent, config);
    const itemType = getItemType(post, parent);
    unirest.post(yonApiUrl)
        .headers({
          'Accept': 'application/json',
          'Content-Type': 'application/json',
          'Yona-Token': config.YONA.TO.USER_TOKEN
        })
        .send(JSON.stringify({ [itemType]: [data] }))
        .end(response => {
          if (isBadResponse(response.status)) {
            console.log('오류 발생!! HTTP 응답코드를 확인하세요! ', yonApiUrl, response.status, response.statusMessage, JSON.stringify(data));
          }
          console.log(response.body);
          this.okCount++;
          if (cb) {
            return cb(response);
          }
        });
  }

  importData(payload, apiUrl, cb) {
    unirest.post(apiUrl)
        .headers({
          'Accept': 'application/json',
          'Content-Type': 'application/json',
          'Yona-Token': config.YONA.TO.USER_TOKEN
        })
        .send(JSON.stringify(payload))
        .end(response => {
          if (isBadResponse(response.status)) {
            console.log('오류 발생!! HTTP 응답코드를 확인하세요! ', apiUrl, response.status, response.statusMessage);
          }
          console.log(response.body);
          if (cb) {
            return cb(response);
          }
        });
  }

  collectAttachmentFileIds(target) {
    let files = [];
    if (target.attachments) {
      target.attachments.forEach(file => {
        files.push(file.uploadedFile);
      });
    }

    console.log('temporary upload files ', files);
    return files;
  }

  pushFiles(post, parent, cb) {
    let uploaded = 0;
    if (!post.attachments || post.attachments.length === 0) {
      this.pushPost(post, parent, cb);
      return;
    }

    let attachments = [...post.attachments];
    let updatedAttachments = [];

    console.log(post.id, ' to upload file count ', attachments.length);
    uploadFiles(attachments, cb);

    function uploadFiles(attachments, cb) {
      if(attachments && attachments.length > 0){
        const file = attachments.pop();
        return upload(file, post.author.loginId, post.author.email, attachment => {
          console.log(++uploaded, 'is done');
          updatedAttachments.push(attachment);
          uploadFiles(attachments, cb);
        })
      }
      post.attachments = updatedAttachments;
      const yonaExport = new YonaExport();
      yonaExport.pushPost(post, parent, cb);
    }

    function upload(attachment, authorLoginId, authorEmail, cb){
      const attachmentBaseDir = path.join(config.EXPORT_BASE_DIR,
          config.YONA.FROM.OWNER_NAME,
          config.YONA.FROM.PROJECT_NAME,
          config.ATTACHMENTS_DIR);

      console.log('to upload file: ', attachment);
      // TODO Check this path is correct
      const filePath = path.join(attachmentBaseDir, attachment.id.toString(), attachment.name);
      if (!fse.existsSync(filePath)) {
        console.error("File not found! - ", __dirname, filePath);
      }

      unirest.post(config.YONA.TO.SERVER + '/files')
          .headers({
            'Accept': 'application/json',
            'Content-Type': 'multipart/form-data',
            'Yona-Token': config.YONA.TO.USER_TOKEN
          })
          .field('authorLoginId', authorLoginId)
          .field('authorEmail', authorEmail)
          .attach('filePath', filePath) // Attachment
          .end(response => {
            if (isBadResponse(response.status)) {
              console.log('Attachment 오류 발생!! HTTP 응답코드를 확인하세요! ',
                  response.status, response.statusMessage, filePath);
            } else {
              attachment.uploadedFile = response.headers.location.split('/')[2];   // eg. location: /files/71
            }
            console.log(response.body);
            cb(attachment);
          });
    }
  }
}

function isBadResponse(statusCode) {
  return [200, 201].indexOf(statusCode) === -1;
}
