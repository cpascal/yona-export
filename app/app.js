import unirest from 'unirest';
import fse from 'fs-extra';
import path from 'path';
import config from '../config';
import { getDefaultFileName } from './utils';
import { createHeader } from './header';
import YonaExport from './YonaExport';

const yonaProjectExportUrl = config.YONA.FROM.SERVER + '/-_-api/v1'
    + '/owners/' + config.YONA.FROM.OWNER_NAME
    + '/projects/' + config.YONA.FROM.PROJECT_NAME
    + '/exports';

const main = () => {
  exportFrom();
};

main();

function exportTo() {
  const from = config.YONA.FROM;
  const to = config.YONA.TO;
  const exportedFile = path.join('exported/',
      from.OWNER_NAME,
      from.PROJECT_NAME + '.json');
  console.log(exportedFile);
  const exportedData = JSON.parse(fse.readFileSync(exportedFile, 'utf8'));

  const project = parseProject(exportedData, to);
  const users = parseRequiredUsers(project);

  const yonaExport = new YonaExport();
  yonaExport.pushToCreateUsers(users, () =>
      yonaExport.pushProject(project, () =>
          yonaExport.pushMilestones({milestones: exportedData.milestones})
      ));

}

function parseRequiredUsers(project) {
  return {users: [...project.members, ...project.authors, ...project.assignees] }

}

function parseProject(source, to) {
  return {
    owner: to.OWNER_NAME,
    projectName: to.PROJECT_NAME,
    projectDescription: source.projectDescription,
    projectCreatedDate: source.projectCreatedDate,
    projectVcs: source.projectVcs,
    projectScope: source.projectScope,
    assignees: source.assignees,
    authors: source.authors,
    members: source.members,
    labels: source.labels
  }
}

function exportFrom() {
  unirest.get(yonaProjectExportUrl)
      .headers({
        'Accept': 'application/json',
        'Yona-Token': config.YONA.FROM.USER_TOKEN
      })
      .end(response => {
        if (isBadResponse(response.status)) {
          console.log('오류 발생!! HTTP 응답코드를 확인하세요! ', yonaProjectExportUrl, response.status, response.statusMessage);
          process.exit(1);
        }
        const exportDir = path.join(
            config.EXPORT_BASE_DIR,
            config.YONA.FROM.OWNER_NAME,
            config.YONA.FROM.PROJECT_NAME,
        );
        writeOriginalJson(response.body, exportDir);
        writeItems(response.body.issues, path.join(exportDir, '/issues/'));
        writeItems(response.body.posts, path.join(exportDir, '/posts/'));
        writeItems(response.body.milestones, path.join(exportDir, '/milestones/'));

        if (!config.EXPORT_ONLY) {
          exportTo();
        }
      });
}



function isBadResponse(statusCode) {
  return [200, 201].indexOf(statusCode) === -1;
}

function writeOriginalJson(item, exportDir){
  fse.outputFile(`${exportDir}.json`, JSON.stringify(item), err => {
    if (err) console.error(err);
  })
}

function writeItems(items, exportDir) {
  items.forEach(item => {
    fse.outputFile(`${exportDir}${getDefaultFileName(item)}`, createHeader(item), err => {
      if (err) console.error(err);
    })
  });
}
