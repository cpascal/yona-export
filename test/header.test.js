import { createHeader } from '../app/header';
import json from './resource/exports';

describe('header', () => {
  it('createHeader', () => {
    // Given // When
    const header = createHeader(json.issues[0]);

    // Then
    expect(header).toMatchSnapshot(); // ITEM_POST
  });
});
