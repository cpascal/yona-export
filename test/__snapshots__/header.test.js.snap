// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header createHeader 1`] = `
"\`\`\`
id: 1
title: 새로운 이슈
type: ISSUE_POST
author: doortts
authorName: 너굴
createdAt: 2016-11-19T08:12:03.000Z
attachments:
  - id: 21957
    name: bensound-cute.mp3
    hash: 0df209c0a171931c4612ac8edefc0d681498148e
    mimeType: audio/mpeg
    size: 2718353
comments:
  - id: 1
    authorId: doortts
    authorName: 너굴
    createdAt: 1479543131000
    body: \\"넹 그럿습니다.\\\\r\\\\nㅋㅋ\\"
  - id: 2
    authorId: doortts
    authorName: 너굴
    createdAt: 1479579108000
    body: \\"정말요?\\\\r\\\\n\\"
  - id: 3
    authorId: doortts
    authorName: 너굴
    createdAt: 1479579129000
    body: \\"넹..ㅋㅋ[bensound-memories.mp3](/files/21959)\\\\r\\\\n[bensound-cute.mp3](/files/21958)\\\\r\\\\n\\"
    attachments:
      - id: 21958
        name: bensound-cute.mp3
        hash: 0df209c0a171931c4612ac8edefc0d681498148e
        mimeType: audio/mpeg
        size: 2718353
      - id: 21959
        name: bensound-memories.mp3
        hash: 7e89fcc98822bdd2c0da64b23178bfb6ca7d78dc
        mimeType: audio/mpeg
        size: 3224136
\`\`\`
이슈인거죠!!
[bensound-cute.mp3](/files/21957)"
`;
